### PR TITLE
Subway station connections

### PR DESCRIPTION
**Pull request by TaliBenedek**

### CODE REVIEW

File: src/main/java/SubwayLines.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Iterator;
+
+public class SubwayLines extends HashMap<String, String[]>
+{
+    public String[] getConnectedStations(String stationName, SubwayStations stations)
+    {
+        String objectId = stations.getObjectId(stationName);
+        ArrayList<String> connections = new ArrayList<>();
+        Iterator<Entry<String, String[]>> it = this.entrySet().iterator();
+        while(it.hasNext())
```
Comment (andrewoid) [inline src/main/java/SubwayLines.java]: Use a for each loop

File: src/main/java/SubwayStations.java [hunk 2]
```diff
@@ -1,9 +1,33 @@
 import java.util.List;
 
-public class Subway
+public class SubwayStations
 {
     List<Feature> features;
 
+    public String getName(String objectId)
+    {
+        for(int index = 0; index < features.size(); index++)
```
Comment (andrewoid) [inline src/main/java/SubwayStations.java]: Use a for each loop

File: src/main/java/SubwayStations.java [hunk 3]
```diff
@@ -1,9 +1,33 @@
...
+            {
+                return this.features.get(index).getName();
+            }
+        }
+        return null;
+    }
+
+    public String getObjectId(String name)
+    {
+        for(int index = 0; index < features.size(); index++)
```
Comment (andrewoid) [inline src/main/java/SubwayStations.java]: Use a for each loop

File: src/main/java/JsonMapReader.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class JsonMapReader
```
Comment (andrewoid) [inline src/main/java/JsonMapReader.java]: This isn't a good name for what the class does.

File: src/main/java/JsonObjectReader.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class JsonObjectReader
```
Comment (andrewoid) [inline src/main/java/JsonObjectReader.java]: Bad name for the class.

File: src/main/java/SubwayStations.java [hunk 6]
```diff
@@ -19,7 +43,7 @@ public String getName()
             return properties.line.split("-");
         }
 
-        public int getObjectId()
+        public String getObjectId()
```
Comment (andrewoid) [inline src/main/java/SubwayStations.java]: Why did you change this from an int to a string?

File: src/test/java/JsonMapReaderTest.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+public class JsonMapReaderTest
+{
+    @Test
+    public void readJsonAsMap() throws IOException
```
Comment (andrewoid) [inline src/test/java/JsonMapReaderTest.java]: What method on JsonMapReader are you testing?

File: src/main/java/SubwayStations.java (hunk 6 above)
Comment (TaliBenedek) [reply src/main/java/SubwayStations.java]: Because when I was coding the getConnectedStations method I saw it would be easier to compare the objectIds of SubwayLines.json and SubwayStations.json if they were both of the same type since then there is no need to cast anything. Then, once I looked at SubwayStations.json again, I saw objectId was really a string, so it was a mistake for me to make it an int to begin with.

File: src/test/java/JsonMapReaderTest.java (hunk 7 above)
Comment (TaliBenedek) [reply src/test/java/JsonMapReaderTest.java]: I'm testing `readJsonAsMap`, but I see I'm really testing `getConnectedStations`, so I really should put that in a separate test for SubwayLines and in this test I should make sure it read the Json into the map correctly with the proper keys and values, right?

File: src/main/java/SubwayStations.java (hunk 6 above)
Comment (andrewoid) [reply src/main/java/SubwayStations.java]: A feature of Gson is that even though it's a String in the json, if it's a number it will automatically convert it for you.

So if you have
int objectid;
It will work.

Comment (TaliBenedek) [reply src/main/java/SubwayStations.java]: In this case though there isn't a reason to make objectId an int since it's just being used as an identifier and not as an actual int, correct?

Comment (andrewoid) [reply src/main/java/SubwayStations.java]: Ints are easier to compare than Strings.

File: src/main/java/JsonToSubwayLines.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class JsonToSubwayLines
+{
+    SubwayLines lines;
```
Comment (andrewoid) [inline src/main/java/JsonToSubwayLines.java]: You don't need this as a member variable if it's just being created and returned in `readJsonAsMap()`

File: src/main/java/JsonToSubwayStations.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class JsonToSubwayStations
+{
+    SubwayStations stations;
```
Comment (andrewoid) [inline src/main/java/JsonToSubwayStations.java]: Same as above.

**Review by andrewoid — APPROVED** [review]

File: src/main/java/SubwayLines.java [hunk 10]
```diff
@@ -4,28 +4,39 @@
 
 public class SubwayLines extends HashMap<String, int[]>
 {
-    public String[] getConnectedStations(String stationName, SubwayStations stations)
+    public ArrayList<SubwayStations.Station> getConnectedStations(SubwayStations subway, SubwayStations.Station station)
     {
-        int objectId = stations.getObjectId(stationName);
-        ArrayList<String> connections = new ArrayList<>();
+        ArrayList<SubwayStations.Station> connections = new ArrayList<>();
+        Map map = subway.getStations();
```
Comment (andrewoid) [inline src/main/java/SubwayLines.java]: `Map<Integer, Station>`

**Review by andrewoid — APPROVED** [review]
